`azurerm_hdinsight_*` - support HDInsight clusters with Data Lake Gen2 Filesystems

dintskirveli opened the pull request:
Allows setting up HDInsight clusters using Data Lake Gen2 Filesystems for storage.

```
  storage_account_gen2 {
    storage_resource_id = azurerm_storage_account.landingarea.id
    filesystem_id  = azurerm_storage_data_lake_gen2_filesystem.landingarea.id
    managed_identity_resource_id = azurerm_user_assigned_identity.my-identity.id
    is_default           = true
  }
```

dintskirveli commented:
@tombuildsstuff any chance we can get a review here?

dintskirveli commented:
@katbyte @tombuildsstuff can we please get a review here?

dintskirveli commented:
> Hi @dintskirveli,
> 
> Thank you for this PR. Looking over the schema i am wondering it might not be better to create a new block `storage_account_gen2` to simplify the logic, and then on expand/flatten we can just merge/split them by the type of storage account?
> 
> Also it looks like you have only updated the expand function. I think you need to update the flatten one too? as well as write some tests to ensure it works with each affected resource.

@katbyte, thanks for looking!

I agree that `storage_account_gen2` might be a better option.

As for tests... I can give them a try, but creating managed identities and assigning roles to them requires "Owner"-level privileges on the testing account. Are the test accounts expected to have those kinds of permissions?

katbyte commented:
@dintskirveli,

For our CI they should have permission, and if not we can see about finagling something so we can run the tests 🙂

dintskirveli commented:
Working on tests + updated documentation.

Based on this comment, there's no `flatten` to implement: https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_hdinsight_hadoop_cluster.go#L249

I don't think it matters since it is not possible update the storage account after creation of the cluster.

dintskirveli commented:
Hi @katbyte! 

This is ready for another pass. I have added tests for all relevant cluster types, split Gen2 config into a separate block, and added documentation. Note that I've omitted a few HDInsight cluster types because they do not support Gen2 storage:
- storm
- rserver
- ml_services

dintskirveli commented on azurerm/resource_arm_hdinsight_hadoop_cluster.go:
```diff
@@ -93,6 +93,8 @@ func resourceArmHDInsightHadoopCluster() *schema.Resource {
 
 			"storage_account": azure.SchemaHDInsightsStorageAccounts(),
 
+			"storage_account_gen2": azure.SchemaHDInsightsGen2StorageAccounts(),
```
You can have both of these, yes

dintskirveli commented on azurerm/helpers/azure/hdinsight.go:
```diff
@@ -179,20 +179,60 @@ func SchemaHDInsightsStorageAccounts() *schema.Schema {
+					Type:         schema.TypeString,
+					Required:     true,
+					ForceNew:     true,
+					ValidateFunc: validate.NoEmptyStrings,
```
Good call!

dintskirveli commented on website/docs/r/hdinsight_hadoop_cluster.html.markdown:
```diff
@@ -172,6 +174,22 @@ A `storage_account` block supports the following:
+
+* `is_default` - (Required) Is this the Default Storage Account for the HDInsight Hadoop Cluster? Changing this forces a new resource to be created.
+
+-> **NOTE:** One of the `storage_account` or `storage_account_gen2` blocks must be marked as the default.
```
Hmm, it looks to me like the convention is to have the NOTE follow the most relevant attribute. Let me know if you do still want this.

dintskirveli commented:
@katbyte 

Thanks for the review!

- I fixed the test failure, it was caused by mishandling of storage_account_gen2 in cluster types that don't support it
- Improved the validation of resource IDs per your suggestion
- Holding off on the documentation change, let me know if you still feel it's necessary.
